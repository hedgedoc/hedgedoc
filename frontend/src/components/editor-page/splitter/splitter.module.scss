/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
.left, .right {
  overflow: hidden;
}

.inner {
  display: flex;
  min-width: 20vw;
  height: 100%;
  flex-direction: column;
}

.resizing {
  .left, .right {
    user-select: none;
  }
}

.move-overlay {
  position: absolute;
  height: 100%;
  width: 100%;
  z-index: 100000;
}
