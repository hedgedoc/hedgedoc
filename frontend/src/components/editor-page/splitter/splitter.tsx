/*
 * SPDX-FileCopyrightText: 2024 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { useKeyboardShortcuts } from './hooks/use-keyboard-shortcuts'
import { DividerButtonsShift, SplitDivider } from './split-divider/split-divider'
import styles from './splitter.module.scss'
import type { MouseEvent, ReactElement, TouchEvent } from 'react'
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { useApplicationState } from '../../../hooks/common/use-application-state'
import { setEditorSplitPosition } from '../../../redux/editor-config/methods'

export interface SplitterProps {
  left?: ReactElement
  right?: ReactElement
  additionalContainerClassName?: string
}

/**
 * Checks if the given {@link Event} is a {@link MouseEvent}.
 *
 * @param event the event to check
 * @return {@link true} if the given event is a {@link MouseEvent}
 */
const isMouseEvent = (event: MouseEvent | TouchEvent): event is MouseEvent => {
  return (event as MouseEvent).buttons !== undefined
}

const isLeftMouseButtonClicked = (mouseEvent: MouseEvent): boolean => {
  return mouseEvent.buttons === 1
}

/**
 * Extracts the absolute horizontal position of the mouse or first touch point from the event.
 *
 * @param moveEvent The mouse or touch event that contains a pointer position
 * @return the extracted horizontal position or {@link undefined} if no position could be extracted
 */
const extractHorizontalPosition = (moveEvent: MouseEvent | TouchEvent): number | undefined => {
  if (isMouseEvent(moveEvent)) {
    return moveEvent.clientX
  } else {
    return moveEvent.touches[0]?.clientX
  }
}

const SNAP_PERCENTAGE = 10

/**
 * Creates a Left/Right splitter react component.
 *
 * @param additionalContainerClassName css classes that are added to the split container.
 * @param left the react component that should be shown on the left side.
 * @param right the react component that should be shown on the right side.
 * @return the created component
 */
export const Splitter: React.FC<SplitterProps> = ({ additionalContainerClassName, left, right }) => {
  const relativeSplitValue = useApplicationState((state) => state.editorConfig.splitPosition)
  const [resizingInProgress, setResizingInProgress] = useState(false)
  const adjustedRelativeSplitValue = useMemo(() => Math.min(100, Math.max(0, relativeSplitValue)), [relativeSplitValue])
  const splitContainer = useRef<HTMLDivElement>(null)

  /**
   * Starts the splitter resizing.
   */
  const onStartResizing = useCallback(() => {
    setResizingInProgress(true)
  }, [])

  /**
   * Stops the splitter resizing.
   */
  const onStopResizing = useCallback(() => {
    setResizingInProgress(false)
  }, [])

  /**
   * Moves the splitter to the left or right side if the relative split value is close to the edges.
   */
  useEffect(() => {
    if (!resizingInProgress) {
      if (relativeSplitValue < SNAP_PERCENTAGE && relativeSplitValue > 0) {
        setEditorSplitPosition(0)
      }
      if (relativeSplitValue > 100 - SNAP_PERCENTAGE && relativeSplitValue < 100) {
        setEditorSplitPosition(100)
      }
    }
  }, [resizingInProgress, relativeSplitValue])

  /**
   * Recalculates the panel split based on the absolute mouse/touch position.
   *
   * @param moveEvent is a {@link MouseEvent} or {@link TouchEvent} that got triggered.
   */
  const onMove = useCallback((moveEvent: MouseEvent | TouchEvent) => {
    if (!splitContainer.current) {
      return
    }
    if (isMouseEvent(moveEvent) && !isLeftMouseButtonClicked(moveEvent)) {
      setResizingInProgress(false)
      moveEvent.preventDefault()
      return undefined
    }

    const horizontalPosition = extractHorizontalPosition(moveEvent)
    if (horizontalPosition === undefined) {
      return
    }
    const horizontalPositionInSplitContainer = horizontalPosition - splitContainer.current.offsetLeft
    const newRelativeSize = horizontalPositionInSplitContainer / splitContainer.current.clientWidth
    const number = newRelativeSize * 100
    setEditorSplitPosition(number)
    moveEvent.preventDefault()
  }, [])

  const onLeftButtonClick = useCallback(() => {
    setEditorSplitPosition(relativeSplitValue === 100 ? 50 : 0)
  }, [relativeSplitValue])

  const onRightButtonClick = useCallback(() => {
    setEditorSplitPosition(relativeSplitValue === 0 ? 50 : 100)
  }, [relativeSplitValue])

  const dividerButtonsShift = useMemo(() => {
    if (relativeSplitValue === 0) {
      return DividerButtonsShift.SHIFT_TO_RIGHT
    } else if (relativeSplitValue === 100) {
      return DividerButtonsShift.SHIFT_TO_LEFT
    } else {
      return DividerButtonsShift.NO_SHIFT
    }
  }, [relativeSplitValue])

  useKeyboardShortcuts()

  return (
    <div
      ref={splitContainer}
      className={`flex-fill flex-row d-flex ${additionalContainerClassName || ''}${
        resizingInProgress ? ' ' + styles.resizing : ''
      }`}>
      {resizingInProgress && (
        <div
          className={styles['move-overlay']}
          onTouchMove={onMove}
          onMouseMove={onMove}
          onTouchCancel={onStopResizing}
          onTouchEnd={onStopResizing}
          onMouseUp={onStopResizing}></div>
      )}
      <div
        id={'editor-edit-pane'}
        className={styles['left']}
        style={{ width: `calc(${adjustedRelativeSplitValue}% - 5px)` }}>
        <div className={styles['inner']}>{left}</div>
      </div>
      <SplitDivider
        onGrab={onStartResizing}
        onLeftButtonClick={onLeftButtonClick}
        onRightButtonClick={onRightButtonClick}
        forceOpen={resizingInProgress}
        focusLeft={relativeSplitValue < SNAP_PERCENTAGE}
        focusRight={relativeSplitValue > 100 - SNAP_PERCENTAGE}
        dividerButtonsShift={dividerButtonsShift}
      />
      <div
        id={'editor-view-pane'}
        className={styles['right']}
        style={{ width: `calc(100% - ${adjustedRelativeSplitValue}%)` }}>
        <div className={styles['inner']}>{right}</div>
      </div>
    </div>
  )
}
