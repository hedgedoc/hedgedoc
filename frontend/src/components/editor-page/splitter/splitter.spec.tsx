/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { Splitter } from './splitter'
import { fireEvent, render, screen } from '@testing-library/react'
import { Mock } from 'ts-mockery'

describe('Splitter', () => {
  describe('resize', () => {
    beforeEach(() => {
      Object.defineProperty(window.HTMLDivElement.prototype, 'clientWidth', { value: 1920 })
      Object.defineProperty(window.HTMLDivElement.prototype, 'offsetLeft', { value: 0 })
    })

    it('can react to shortcuts', () => {
      const view = render(<Splitter left={<>left</>} right={<>right</>} />)
      fireEvent.keyDown(document, Mock.of<KeyboardEvent>({ ctrlKey: true, altKey: true, key: 'v' }))
      expect(view.container).toMatchSnapshot()
      fireEvent.keyDown(document, Mock.of<KeyboardEvent>({ ctrlKey: true, altKey: true, key: 'e' }))
      expect(view.container).toMatchSnapshot()
      fireEvent.keyDown(document, Mock.of<KeyboardEvent>({ ctrlKey: true, altKey: true, key: 'b' }))
      expect(view.container).toMatchSnapshot()
    })

    it('can change size with mouse', async () => {
      const view = render(<Splitter left={<>left</>} right={<>right</>} />)
      expect(view.container).toMatchSnapshot()
      const divider = await screen.findByTestId('splitter-divider')

      fireEvent.mouseDown(divider, {})
      fireEvent.mouseMove(window, Mock.of<MouseEvent>({ buttons: 1, clientX: 1920 }))
      fireEvent.mouseUp(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.mouseDown(divider, {})
      fireEvent.mouseMove(window, Mock.of<MouseEvent>({ buttons: 1, clientX: 0 }))
      fireEvent.mouseUp(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.mouseMove(window, Mock.of<MouseEvent>({ buttons: 1, clientX: 1920 }))
      expect(view.container).toMatchSnapshot()
    })

    it('can change size with touch', async () => {
      const view = render(<Splitter left={<>left</>} right={<>right</>} />)
      expect(view.container).toMatchSnapshot()
      const divider = await screen.findByTestId('splitter-divider')

      fireEvent.touchStart(divider, {})
      fireEvent.touchMove(window, Mock.of<TouchEvent>({ touches: [{ clientX: 1920 }, { clientX: 200 }] }))
      fireEvent.touchEnd(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.touchStart(divider, {})
      fireEvent.touchMove(window, Mock.of<TouchEvent>({ touches: [{ clientX: 0 }, { clientX: 100 }] }))
      fireEvent.touchCancel(window)
      expect(view.container).toMatchSnapshot()

      fireEvent.touchMove(window, Mock.of<TouchEvent>({ touches: [{ clientX: 500 }, { clientX: 900 }] }))
      expect(view.container).toMatchSnapshot()
    })
  })
})
