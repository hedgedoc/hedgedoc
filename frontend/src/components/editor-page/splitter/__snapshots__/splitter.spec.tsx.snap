// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splitter resize can change size with mouse 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 2`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 3`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with mouse 4`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 2`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 3`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can change size with touch 4`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can react to shortcuts 1`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(0% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle shift-right"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 0%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can react to shortcuts 2`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(100% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle shift-left"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-secondary"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 100%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;

exports[`Splitter resize can react to shortcuts 3`] = `
<div>
  <div
    class="flex-fill flex-row d-flex "
  >
    <div
      class="left"
      style="width: calc(50% - 5px);"
    >
      <div
        class="inner"
      >
        left
      </div>
    </div>
    <div
      class="divider"
      data-testid="splitter-divider"
    >
      <div
        class="middle"
      >
        <div
          class="buttons"
        >
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowLeft
          </button>
          <span
            class="grabber"
          >
            BootstrapIconMock_ArrowLeftRight
          </span>
          <button
            class="btn btn-light"
            type="button"
          >
            BootstrapIconMock_ArrowRight
          </button>
        </div>
      </div>
    </div>
    <div
      class="right"
      style="width: calc(100% - 50%);"
    >
      <div
        class="inner"
      >
        right
      </div>
    </div>
  </div>
</div>
`;
